pkg/crdschema/gen.go

```diff
@@ -0,0 +1,377 @@
+		for _, s := range versionSchemas {
+			if !reflect.DeepEqual(s, first) {
+				allEqual = false
+				break
+			}
+		}
+
+		// delete old and set new schemas
+		var changed bool
+		if allEqual {
+			changed, err = existing.setGlobalSchema(first)
+			if err != nil {
+				return fmt.Errorf("failed to set global schema for %s in %s: %v", groupKind.Kind, groupKind.Group, err)
+			}
+		} else {
+			changed, err = existing.setVersionedSchema(versionSchemas)
+			if err != nil {
+				return fmt.Errorf("failed to set versioned schemas for %s in %s: %v", groupKind.Kind, groupKind.Group, err)
+			}
+		}
+		if changed {
+			fmt.Printf("Updating schema for %s in %s at spec.validation.openAPIV3Schema.\n", groupKind.Kind, groupKind.Group)
+		} else {
+			fmt.Printf("Unchanged schema for %s in %s at spec.validation.openAPIV3Schema.\n", groupKind.Kind, groupKind.Group)
+		}
+
+		// check whether the file changes
+		if changed, err := existing.compareToExistingFile(ctx, fn); err != nil {
+			return fmt.Errorf("failed to compare against %q: %v", fn, err)
+		} else if changed {
+			changedFiles = append(changedFiles, fn)
+		}
+	}
+
+	// only check for differences in verify-only mode
+	if g.VerifyOnly {
+		if len(changedFiles) > 0 {
+			return fmt.Errorf("unexpected changes to: %s", strings.Join(changedFiles, ", "))
+		}
+		return nil
+	}
+
+	return g.writeChangedFiles(ctx, changedFiles, existingCRDsByFileNames)
+}
+
+func (g Generator) writeChangedFiles(ctx *genall.GenerationContext, changedFiles []string, existingCRDsByFileNames map[string]*existingCRD) (result error) {
+	// write changes
+	for _, fn := range changedFiles {
+		bs, err := yaml.Marshal(existingCRDsByFileNames[fn].Yaml)
+		if err != nil {
+			return err
+		}
+		if ctx.OutputRule == genall.OutputToNothing {
+			fmt.Printf("Not writing changed %s.\n", fn)
+			continue
+		}
+
+		rel, err := filepath.Rel(g.ManifestsPath, fn)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("Writing %s via output rule.\n", rel)
+		w, err := ctx.OutputRule.Open(nil, rel)
+		if err != nil {
+			return err
+		}
+		if _, err := w.Write(bs); err != nil {
+			return err
+		}
+		if err := w.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type existingCRD struct {
+	GroupKind schema.GroupKind
+	Yaml      interface{}
+	CRD       *apiextensionsv1beta1.CustomResourceDefinition
+}
+
+func (e *existingCRD) compareToExistingFile(ctx *genall.GenerationContext, fn string) (bool, error) {
+	bs, err := ctx.ReadFile(fn)
+	if err != nil {
+		return false, err
+	}
+	var old yaml.MapSlice
+	if err := yaml.Unmarshal(bs, &old); err != nil {
+		return false, err
+	}
+	y := e.Yaml.(yaml.MapSlice)
+	return !reflect.DeepEqual(old, y), nil
+}
+
+func (e *existingCRD) setGlobalSchema(schema *apiextensionsv1beta1.JSONSchemaProps) (bool, error) {
+	y, err := crdschemayaml.ToYAML(schema)
+	if err != nil {
+		return false, err
+	}
+	old, _, err := crdschemayaml.GetNestedFieldNoCopy(e.Yaml, "spec", "validation", "openAPIV3Schema")
```

Comment:
can just use unstructured helpers here -- no need to copy these in

Reply:
or, potentially even better, just have the jsonpatch diff lib produce a patch and then apply it -- then you get structure.

Reply:
nope. For the same reason as before: we can't destroy order of given YAMLs. Patch would do that.

Reply:
Also note that we agreed on this at KubeCon. Those functions have tests now. And they are in an internal package to not leak anywhere.